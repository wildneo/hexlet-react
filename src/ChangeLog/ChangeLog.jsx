import React from 'react';
import uniqueId from 'lodash.uniqueid';

export default class ChangeLog extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      current: 0,
      log: [],
    };

    this.increment = (e) => {
      e.preventDefault();
      this.pushToLog(1);
    };

    this.decrement = (e) => {
      e.preventDefault();
      this.pushToLog(-1);
    };

    this.removeLogItem = (id) => (e) => {
      e.preventDefault();
      const { log } = this.state;
      const newLog = log.filter((item) => item.id !== id);
      this.setState({ log: newLog });
    };
  }

  pushToLog(value) {
    const { current, log } = this.state;
    const id = uniqueId();
    const newValue = log.length > 0 ? current + value : 0 + value;
    const newItem = { id, value: newValue };
    this.setState({
      current: newValue,
      log: [newItem, ...log],
    });
  }

  renderLog() {
    const { log } = this.state;

    return (
      <div className="list-group">
        {log.map(({ id, value }) => (
          <button key={id} onClick={this.removeLogItem(id)} type="button" className="list-group-item list-group-item-action">
            {value}
          </button>
        ))}
      </div>
    );
  }

  render() {
    const { log } = this.state;

    return (
      <div>
        <div className="btn-group" role="group">
          <button type="button" className="btn hexlet-inc" onClick={this.increment}>+</button>
          <button type="button" className="btn hexlet-dec" onClick={this.decrement}>-</button>
        </div>
        { log.length > 0 ? this.renderLog() : null }
      </div>
    );
  }
}
