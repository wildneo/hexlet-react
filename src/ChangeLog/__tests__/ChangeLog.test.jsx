import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import ChangeLog from '../ChangeLog';

Enzyme.configure({ adapter: new Adapter() });

test('ChangeLog', () => {
  const wrapper = mount(<ChangeLog />);
  expect(wrapper.render()).toMatchSnapshot();

  const inc = wrapper.find('.hexlet-inc');
  const dec = wrapper.find('.hexlet-dec');

  inc.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  inc.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  inc.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const items1 = wrapper.find('.list-group-item-action');
  items1.at(1).simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  const items2 = wrapper.find('.list-group-item-action');
  items2.at(1).simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  inc.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();

  dec.simulate('click');
  expect(wrapper.render()).toMatchSnapshot();
});
