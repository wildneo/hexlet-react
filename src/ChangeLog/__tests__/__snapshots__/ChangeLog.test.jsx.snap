// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChangeLog 1`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
</div>
`;

exports[`ChangeLog 2`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 3`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 4`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 5`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 6`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 7`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 8`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      0
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 9`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      0
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 10`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      0
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;

exports[`ChangeLog 11`] = `
<div>
  <div
    class="btn-group"
    role="group"
  >
    <button
      class="btn hexlet-inc"
      type="button"
    >
      +
    </button>
    <button
      class="btn hexlet-dec"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="list-group"
  >
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      0
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      0
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      2
    </button>
    <button
      class="list-group-item list-group-item-action"
      type="button"
    >
      1
    </button>
  </div>
</div>
`;
